File: app/iOS/Thermal Comfort Study/SurveyDataItem.swift
//
//  SurveyDataItem.swift
//  Thermal Comfort Study
//
//  Created by Madhav Achar on 10/21/16.
//  Copyright © 2016 boschsc. All rights reserved.
//

import Foundation

public class SurveyDataItem: NSObject {
    
    public static let ITEM_ID_PREFIX = "_SurveyData"
    
    private var username: String!
    private var dataName: String!
    private var value: String!
    private var timestamp: String!
    
    init(_ username: String, _ timestamp: Date, _ dataName: String, _ value: String) {
        self.username = username
        self.dataName = dataName
        self.value = value
        
        let dateFormatter = DateFormatter()
        dateFormatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ss.SSSSSSZ"
        self.timestamp = dateFormatter.string(from: timestamp)
    }
    
    private func getID() -> String {
        return SurveyDataItem.ITEM_ID_PREFIX + timestamp
    }
    
    public func toXML() -> String {
        var str = ""
        str.append("<item id=\"" + self.getID() + "\">")
        str.append("<transducerData id=\"occthermcomf_bosch\" ")
        str.append("type=\"TCS\" ")
        str.append("class=\"SurveyData\" ")
        str.append("username=\"\(username!)\" ")
        str.append("timestamp=\"\(timestamp!)\" ")
        str.append("name=\"\(dataName!)\" ")
        str.append("value=\"\(value!)\"/>")
        str.append("</item>")
        
        return str
        
    }
    
    public func set(timestamp: Date) {
        let dateFormatter = DateFormatter()
        dateFormatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ss.SSSSSSZ"
        self.timestamp = dateFormatter.string(from: timestamp)
    }
}

File: app/android/app/src/main/java/bosch/smartcampus/thermalcomfortstudy/packet/SurveyDataItem.java
package bosch.smartcampus.thermalcomfortstudy.packet;

import org.jivesoftware.smackx.pubsub.Item;

import bosch.smartcampus.thermalcomfortstudy.lib.Timestamp;

/**
 * Created by rsukkerd on 6/14/16.
 */
public class SurveyDataItem extends Item {
    public static final String NAMESPACE = null;
    public static final String ELEMENT = "item";
    public static final String ITEM_ID_PREFIX = "_SurveyData";

    public static final String TRANSDUCER_DATA_ID = "occthermcomf_bosch";
    public static final String TRANSDUCER_DATA_TYPE = "TCS";
    public static final String TRANSDUCER_DATA_CLASS = "SurveyData";

    private String username;
    private Timestamp timestamp;

    public SurveyDataItem(String username, Timestamp timestamp) {
        this.username = username;
        this.timestamp = timestamp;
    }

    public String getUsername() {
        return username;
    }

    public Timestamp getTimestamp() {
        return timestamp;
    }

    @Override
    public String getNamespace() {
        return NAMESPACE;
    }

    @Override
    public String getElementName() {
        return ELEMENT;
    }

    @Override
    public String getId() {
        return ITEM_ID_PREFIX + timestamp;
    }

    @Override
    public String toXML() {
        StringBuilder builder = new StringBuilder();
        builder.append("<" + ELEMENT + " id=\"" + getId() + "\">");
        builder.append("</" + ELEMENT + ">");
        return builder.toString();
    }

    protected String getTransducerDataElement(String name, String value) {
        String element = String.format("<transducerData id=\"%s\" type=\"%s\" class=\"%s\" ",
                TRANSDUCER_DATA_ID, TRANSDUCER_DATA_TYPE, TRANSDUCER_DATA_CLASS);
        element += "username=\"" + username + "\" ";
        element += "timestamp=\"" + timestamp + "\" ";
        element += "name=\"" + name + "\" ";
        element += "value=\"" + value + "\"/>";
        return element;
    }
}
